framework/include/base/DependencyResolverInterface.h

```diff
@@ -15,49 +15,100 @@
 {
 public:
+  /**
+   * Constructor.
+   */
   DependencyResolverInterface() {}
 
+  /**
+   * Return a set containing the names of items requested by the object.
+   */
   virtual const std::set<std::string> & getRequestedItems() = 0;
+
+  /**
+   * Return a set containing the names of items owned by the object.
+   */
   virtual const std::set<std::string> & getSuppliedItems() = 0;
 
-  template<typename Iter>
+  /**
+   * Given a vector, sort using the getRequested/SuppliedItems sets.
+   */
+  template<typename T>
+  static
+  void sort(typename std::vector<T> & vector);
+
+  /**
+   * A helper method for cyclic errors.
+   * @see MaterialWarehouse
+   */
+  template<typename T>
   static
-  void sort(Iter start, Iter end)
+  void cyclicDependencyError(CyclicDependencyException<T> & e, const std::string & header);
+};
+
+
+template<typename T>
+void
+DependencyResolverInterface::sort(typename std::vector<T> & vector)
+{
+  DependencyResolver<T> resolver;
+
+  typename std::vector<T>::iterator start = vector.begin();
+  typename std::vector<T>::iterator end = vector.end();
+
+  for (typename std::vector<T>::iterator iter = start; iter != end ; ++iter)
   {
-    DependencyResolver<DependencyResolverInterface *> resolver;
+    const std::set<std::string> & requested_items = (*iter)->getRequestedItems();
 
-    for (Iter iter = start; iter != end ; ++iter)
+    for (typename std::vector<T>::iterator iter2 = start; iter2 != end; ++iter2)
     {
-      const std::set<std::string> & requested_items = (*iter)->getRequestedItems();
+      if (iter == iter2)
+        continue;
 
-      for (Iter iter2 = start; iter2 != end; ++iter2)
-      {
-        if (iter == iter2) continue;
-        const std::set<std::string> & supplied_items = (*iter2)->getSuppliedItems();
+      const std::set<std::string> & supplied_items = (*iter2)->getSuppliedItems();
 
-        std::set<std::string> intersect;
-        std::set_intersection(requested_items.begin(), requested_items.end(), supplied_items.begin(),
-                              supplied_items.end(), std::inserter(intersect, intersect.end()));
+      std::set<std::string> intersect;
+      std::set_intersection(requested_items.begin(), requested_items.end(), supplied_items.begin(),
+                            supplied_items.end(), std::inserter(intersect, intersect.end()));
 
-        // If the intersection isn't empty then there is a dependency here
-        if (!intersect.empty())
-          resolver.insertDependency(*iter, *iter2);
-      }
+      // If the intersection isn't empty then there is a dependency here
+      if (!intersect.empty())
+        resolver.insertDependency(*iter, *iter2);
     }
+  }
 
     // Sort based on dependencies
-    std::stable_sort(start, end, resolver);
-  }
-};
+  std::stable_sort(start, end, resolver);
+}
+
+
+template<typename T>
+void
+DependencyResolverInterface::cyclicDependencyError(CyclicDependencyException<T> & e, const std::string & header)
```

Comment:
This is a nice improvement. Don't know why I never bothered to fact this out before.